connection_tracker/bpf/conn_track.c

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+#include "bpf_endian.h"
+#include "bpf_helpers.h"
+#include "net/ip.h"
+#include "..\conn_track\conn_tracker.h"
+
+inline __attribute__((always_inline)) uint64_t
+bpf_ntohll(uint64_t x)
+{
+    uint64_t upper = bpf_ntohl(x >> 32);
+    uint64_t lower = (uint64_t)bpf_ntohl(x & 0xFFFFFFFF) << 32;
+    return upper | lower;
+}
+
+#define bpf_htonll(x) bpf_ntohll(x)
+
+#ifndef ntohll
+#define ntohll bpf_ntohll
+#endif
+#ifndef htonll
+#define htonll bpf_htonll
+#endif
+
+typedef struct _connection_history
+{
+    connection_tuple_t tuple;
+    bool ipv4;
+    uint64_t start_time;
+    uint64_t end_time;
+} connection_history_t;
+
+// Key is the connection tuple, value is the connection start time.
+SEC("maps")
+struct bpf_map_def connection_map = {
+    .type = BPF_MAP_TYPE_LRU_HASH,
+    .key_size = sizeof(connection_tuple_t),
+    .value_size = sizeof(uint64_t),
+    .max_entries = 1024};
+
+// Connection history map containing tuple and connection duration.
+SEC("maps")
+struct bpf_map_def history_map = {.type = BPF_MAP_TYPE_RINGBUF, .max_entries = 256 * 1024};
+
+__attribute__((always_inline)) void
+log_tuple(connection_tuple_t* tuple, bool ipv4, bool connect)
+{
+    if (ipv4) {
+        if (connect) {
+            bpf_printk("Connnection to %x started on if %ld", ntohl(tuple->dst_ip.ipv4), tuple->interface_luid);
+        } else {
+            bpf_printk("Connnection to %x stopped on if %ld", ntohl(tuple->dst_ip.ipv4), tuple->interface_luid);
+        }
+    } else {
+        uint64_t* ip = (uint64_t*)tuple->dst_ip.ipv6;
+        if (connect) {
+            bpf_printk("Connnection to %llx %llx started on if %ld", ntohll(ip[0]), ntohll(ip[1]), tuple->interface_luid);
+        } else {
+            bpf_printk("Connnection to %llx %llx stopped on if %ld", ntohll(ip[0]), ntohll(ip[1]), tuple->interface_luid);
+        }
+    }
+}
+
+__attribute__((always_inline)) void
+sock_ops_to_connection_tuple(bpf_sock_ops_t* ctx, bool ipv4, connection_tuple_t* tuple)
+{
+    tuple->compartment_id = ctx->compartment_id;
+    tuple->interface_luid = ctx->interface_luid;
+
+    if (ipv4) {
+        tuple->src_ip.ipv4 = ctx->local_ip4;
+        tuple->src_port = ctx->local_port;
+        tuple->dst_ip.ipv4 = ctx->remote_ip4;
+        tuple->dst_port = ctx->remote_port;
+        tuple->protocol = ctx->protocol;
```

Comment:
nit (non blocking, ok to defer if you prefer):  it's probably better to move ip-version-agnostic fields outside the if/else to be common between them, i.e. src_port, dst_port, and protocol look to be movable just like compartment_id and interface_luid are outside the if.